### DISCUSSION TITLE
Booting MartyPC with 16K of RAM

### DISCUSSION THREAD

**Discussion by dbalsom** [dbalsom/marty#126] [Q&A]
Here's some information on the 5150's dip switches, courtesy of minuszerodegrees:

![image](https://github.com/dbalsom/martypc/assets/7229541/41112e8a-a7d2-491e-8310-a5994114deaf)

https://www.minuszerodegrees.net/5150/ram/5150_ram_16_64_SW2.jpg

On the 5150, MartyPC will take the specified conventional memory size from the machine configuration and "round up" to the nearest valid DIP setting. Specifying 0x04000 bytes of RAM produces a DIP switch of 00011111, which looks to match the diagram.

Note that we cannot use the latest 1982 BIOS if we want to have less than 64K of RAM - that version of the BIOS assumes that all RAM banks are populated.

Even so, if we attempt to boot using the '81 BIOS and 0x4000 bytes of memory, we get an error:

![screenshot234](https://github.com/dbalsom/martypc/assets/7229541/5e29f34c-6954-49c0-8670-a482a21ff1ae)

**Comment by dbalsom** [comment]
If we refer here: https://www.minuszerodegrees.net/5150/ram/5150_ram_16_64.htm

The first two characters of the error code `04` indicate in which 4k block the RAM error occurred, however the first 16K of memory comprises blocks 0-3 of motherboard bank 0.   Block 4 is in bank 1, and shouldn't be used if we only have 16K of memory!

The enabled banks of memory are controlled by switch block S1. This implies that we are not setting switches 3 and 4 correctly. With 16K of RAM, switch 3 and switch 4 should be ON.

debug output shows us how S1 is set:
```
[2024-06-22T17:54:39Z DEBUG marty_core::devices::ppi] DIP SW1: 10011000
```
Switch #3 is zero! Why?

```
pub const SW1_RAM_BANKS_1: u8 = 0b0001_1000;
pub const SW1_RAM_BANKS_2: u8 = 0b0001_0000;
pub const SW1_RAM_BANKS_3: u8 = 0b0000_1000;
pub const SW1_RAM_BANKS_4: u8 = 0b0000_0000;
```
This is a simple off-by-one error.  The switches are numbered 1-indexed, but I calculated the bits as if they were 0-indexed. So all the bank masks are shifted one too far to the left.

**Comment by dbalsom** [comment]
Fixing that, we can get into ROM BASIC.

![screenshot236](https://github.com/dbalsom/martypc/assets/7229541/4e1ed844-bcba-4364-8e1c-829e665dbea4)

I was not able to get DOS 1.0 to boot with only 16K of RAM. It seems to attempt to execute at address 0; which is where the IVT lives, and gets stuck.

**Reply by Torinde** [reply]
The IBM BIOS loads the bootsector always at 0x7C00 (~32KB) and since IBM PC DIP switches are used only to _inform_ BIOS about the memory size, but _don't actually disable_ the RAM banks - it's possible to boot PC DOS 1.0 in a [quasi-16KB/32KB](https://github.com/joncampbell123/dosbox-x/issues/3664#issuecomment-2182430832) mode, where two banks for a total of 32KB RAM are populated, but DOS/BIOS uses only 16KB.

Can you add a setting for "memory banks actual population", separate from "memory size selected by DIPs"? To replicate [this behavior](https://www.minuszerodegrees.net/5150/ram/5150_not_enough_ram_for_DOS.htm):
- [ ] 16KB DIP, only first bank populated: 'PARITY CHECK 1' on-screen, plus floppy drive access LED on continuously, plus floppy drive spindle turning continuously.
- [ ] 16KB DIP, first & second banks populated: There was a successful boot if all I did was simply populate the second RAM bank, leaving the motherboard switches set to 16 KB.
- [ ] All four banks are _always_ accessible, regardless of DIP switches: [diagram of permanently connected U12 memory transievers](https://www.minuszerodegrees.net/5150/motherboard/IBM%205150%20motherboard%20-%20Removing%20RAM%20chips.png)
So, while BIOS/DOS limit should be decided by the DIP switches, the ability for memory access by programs should be linked to "memory bank actual population".
And when attempting to [read from an empty bank](https://github.com/640-KB/GLaBIOS/discussions/50#discussioncomment-9848064) should get either only ones or only zeros or a random combination of ones and zeroes (regardless of previous write attempts)?
- [ ] authentic, but somewhat superfluous feature will be to simulate faults in memory chips - both for the regular range (per DIPs) and for the extra range (additional populated banks). Do you support that for the regular range already, e.g. simulate faults to see if BIOS memory check routines work as expected?

**Reply by dbalsom** [reply]
> The IBM BIOS loads the bootsector always at 0x7C00 (~32KB) and since IBM PC DIP switches are used only to _inform_ BIOS about the memory size, but _don't actually disable_ the RAM banks - it's possible to boot PC DOS 1.0 in a [quasi-16KB/32KB](https://github.com/joncampbell123/dosbox-x/issues/3664#issuecomment-2182430832) mode, where two banks for a total of 32KB RAM are populated, but DOS/BIOS uses only 16KB.
> 
> Can you add a setting for "memory banks actual population", separate from "memory size selected by DIPs"? To replicate [this behavior](https://www.minuszerodegrees.net/5150/ram/5150_not_enough_ram_for_DOS.htm):

Well, what I would probably implement is a DIP override, so the conventional memory size would be the actual memory you had in the system, but you could set the DIP switches however you felt like it instead of having MartyPC try to figure them out for you. That would let you do this sort of thing.

As for simulating memory errors - well, you saw that in action just a few posts ago. We got an error 201 which is a memory failure, because MartyPC returns FF on an open bus. As far as simulating parity errors, I don't really have any plans, although that would be neat. That's the kind of thing ideally someone contributes via PR so I can concentrate on bigger features.

**Reply by Torinde** [reply]
> implement is a DIP override

Good! I also wonder if all of the second bank needs to be filled - or a single RAM chip at the top of 32KB is sufficient.
- DIP switches at 16KB
- RAM chips from 0x0000-0x3FFF and 0x7800-0x7FFF

**Comment by dbalsom** [comment]
DOS 1.10 boots with 32K of RAM:
![screenshot237](https://github.com/dbalsom/martypc/assets/7229541/db8daade-769e-4127-bff7-613576dfcbba)

**Comment by Torinde** [comment]
GLaBIOS also supports [less than 64KB](https://github.com/640-KB/GLaBIOS/discussions/50#discussioncomment-9847336).
